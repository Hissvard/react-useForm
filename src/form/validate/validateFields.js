import { Map } from 'immutable'

const validateItems = (items, getAllValues) => {
  return items.map(item => {
    return validateMap(item.get('fields'), getAllValues)
  })
}

const validateMap = (fields, getAllValues) => {
  return fields.entrySeq().reduce((acc, [key, field]) => {
    const value = field.getIn(['current', 'value'])
    const optional = field.getIn(['initial', 'optional'], false)
    const { validate, requiredMessage, name } = field.getIn(['initial', 'field'], {})

    const items = field.get('items')
    if (items) {
      return acc.set(key, validateItems(items), getAllValues)
    }

    if (validate) {
      const error = validate(value, name, getAllValues)
      if (error) {
        return acc.set(key, error)
      }
    }

    if (!optional && value !== false && !value) {
      return acc.set(key, requiredMessage || 'Required')
    }

    return acc
  }, Map()).toJS()
}

export const validateFields = (state) => {
  const getAllValues = state.get('getAllValues')
  return validateMap(state.get('fields'), getAllValues)
}
