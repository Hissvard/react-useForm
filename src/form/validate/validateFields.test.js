import { fromJS, Map } from 'immutable'
import { validateFields } from './validateFields'

let state, nicknameField, colorsField, colorField, getAllValues

beforeEach(() => {
  getAllValues = 'getAllValuesFunc'

  nicknameField = Map([
    ['optional', false],
    ['field', {
      name: 'nickname',
      validate: jest.fn(),
    }],
  ])
  colorsField = Map([
    ['optional', false],
    ['field', {
      name: 'colors',
      validate: jest.fn(),
    }],
  ])
  colorField = Map([
    ['optional', false],
    ['field', {
      name: 'color',
      validate: jest.fn(),
    }],
  ])

  state = fromJS({
    other: 'other data',
    getAllValues: 'getAllValuesFunc',
    fields: {
      nickname: {
        current: {
          error: 'notSet',
          helperText: 'notSet',
        },
      },
      colors: {
        current: {
          error: 'notSet',
          helperText: 'notSet',
        },
        items: [
          {
            fields: {
              color: {
                current: {
                  error: 'notSet',
                  helperText: 'notSet',
                },
              },
            },
          },
          {
            fields: {
              color: {
                current: {
                  error: 'notSet',
                  helperText: 'notSet',
                },
              },
              shades: {
                current: {
                  error: 'notSet',
                  helperText: 'notSet',
                },
                items: [
                  {
                    fields: {
                      shade: {
                        current: {
                          error: 'notSet',
                          helperText: 'notSet',
                        },
                      },
                    },
                  },
                ],
              },
            },
          },
        ],
      },
    },
  })

  state = state.setIn(['fields', 'nickname', 'initial'], nicknameField)
  state = state.setIn(['fields', 'colors', 'initial'], colorsField)
  state = state.setIn(['fields', 'colors', 'items', 0, 'fields', 'color', 'initial'], colorField)
  state = state.setIn(['fields', 'colors', 'items', 1, 'fields', 'shades', 'items', 0, 'fields', 'shade', 'initial'], colorField)
})


const setValues = () => {
  state = state
    .setIn(['fields', 'nickname', 'current', 'value'], 'val')
    .setIn(['fields', 'colors', 'items', 0, 'fields', 'color', 'current', 'value'], 'val')
    .setIn(['fields', 'colors', 'items', 1, 'fields', 'color', 'current', 'value'], 'val')
    .setIn(['fields', 'colors', 'items', 1, 'fields', 'shades', 'items', 0, 'fields', 'shade', 'current', 'value'], 'val')
}

test('state', () => {
  setValues()
  expect(state).toMatchSnapshot()
})

test('validates required fields', () => {
  expect(validateFields(state, getAllValues)).toMatchSnapshot()
})

test('validates optional fields', () => {
  setValues()
  expect(validateFields(state, getAllValues)).toMatchSnapshot()
})

test('validates optional = true', () => {
  nicknameField = nicknameField.set('optional', true)
  state = state.setIn(['fields', 'nickname', 'initial'], nicknameField)

  expect(validateFields(state, getAllValues)).toMatchSnapshot()
})

test('validate custom validator', () => {
  setValues()

  const nickname = nicknameField.get('field')
  nickname.validate.mockReturnValue('nickname error')

  expect(validateFields(state, getAllValues)).toMatchSnapshot()
  expect(nickname.validate.mock.calls[0]).toMatchSnapshot()
})
