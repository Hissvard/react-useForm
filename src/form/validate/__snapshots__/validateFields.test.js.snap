// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`state 1`] = `
Immutable.Map {
  "other": "other data",
  "getAllValues": "getAllValuesFunc",
  "fields": Immutable.Map {
    "nickname": Immutable.Map {
      "current": Immutable.Map {
        "error": "notSet",
        "helperText": "notSet",
        "value": "val",
      },
      "initial": Immutable.Map {
        "optional": false,
        "field": Object {
          "name": "nickname",
          "validate": [MockFunction],
        },
      },
    },
    "colors": Immutable.Map {
      "current": Immutable.Map {
        "error": "notSet",
        "helperText": "notSet",
      },
      "items": Immutable.List [
        Immutable.Map {
          "fields": Immutable.Map {
            "color": Immutable.Map {
              "current": Immutable.Map {
                "error": "notSet",
                "helperText": "notSet",
                "value": "val",
              },
              "initial": Immutable.Map {
                "optional": false,
                "field": Object {
                  "name": "color",
                  "validate": [MockFunction],
                },
              },
            },
          },
        },
        Immutable.Map {
          "fields": Immutable.Map {
            "color": Immutable.Map {
              "current": Immutable.Map {
                "error": "notSet",
                "helperText": "notSet",
                "value": "val",
              },
            },
            "shades": Immutable.Map {
              "current": Immutable.Map {
                "error": "notSet",
                "helperText": "notSet",
              },
              "items": Immutable.List [
                Immutable.Map {
                  "fields": Immutable.Map {
                    "shade": Immutable.Map {
                      "current": Immutable.Map {
                        "error": "notSet",
                        "helperText": "notSet",
                        "value": "val",
                      },
                      "initial": Immutable.Map {
                        "optional": false,
                        "field": Object {
                          "name": "color",
                          "validate": [MockFunction],
                        },
                      },
                    },
                  },
                },
              ],
            },
          },
        },
      ],
      "initial": Immutable.Map {
        "optional": false,
        "field": Object {
          "name": "colors",
          "validate": [MockFunction],
        },
      },
    },
  },
}
`;

exports[`validate custom validator 1`] = `
Object {
  "colors": Array [
    Object {},
    Object {
      "shades": Array [
        Object {},
      ],
    },
  ],
  "nickname": "nickname error",
}
`;

exports[`validate custom validator 2`] = `
Array [
  "val",
  "nickname",
  "getAllValuesFunc",
]
`;

exports[`validate custom validator on list 1`] = `
Object {
  "colors": Array [
    Object {},
    Object {
      "shades": Array [
        Object {},
      ],
    },
  ],
}
`;

exports[`validate custom validator on list 2`] = `
Array [
  Immutable.List [
    Immutable.Map {
      "color": "val",
    },
    Immutable.Map {
      "color": "val",
      "shades": undefined,
    },
  ],
  "colors",
  "getAllValuesFunc",
]
`;

exports[`validates optional = true 1`] = `
Object {
  "colors": Array [
    Object {
      "color": "Required",
    },
    Object {
      "color": "Required",
      "shades": Array [
        Object {
          "shade": "Required",
        },
      ],
    },
  ],
}
`;

exports[`validates optional fields 1`] = `
Object {
  "colors": Array [
    Object {},
    Object {
      "shades": Array [
        Object {},
      ],
    },
  ],
}
`;

exports[`validates required fields 1`] = `
Object {
  "colors": Array [
    Object {
      "color": "Required",
    },
    Object {
      "color": "Required",
      "shades": Array [
        Object {
          "shade": "Required",
        },
      ],
    },
  ],
  "nickname": "Required",
}
`;
